backend/auth/role_map_repository_test.go

```diff
@@ -223,3 +239,541 @@ func TestHasPermission(t *testing.T) {
+			log.Printf("Role map created")
+			assert.NotNil(t, repo)
+
+			// Check if the flattened map is created correctly
+			repo.flattenedMap = createPermissionMatrix(repo.RoleMap, repo.SubroleMap)
+			repo.flattenedCompiled = true
+			log.Printf("Flattened map created")
+
+			allPassed := true
+			// Compare hasPermission with flatHasPermission
+			for i := 0; i < tt.namespaces + 1; i++ {
+				for j := 0; j < tt.resources + 1; j++ {
+					for _, op := range operations {
+						operation := &models.Operation{Type: op, Resource: fmt.Sprintf("r%d", j), Namespace: fmt.Sprintf("n%d", i)}
+						expected := hasPermission(repo.RoleMap["root"], repo.SubroleMap, operation, make(map[string]struct{}))
+						result := flatHasPermission(operation, repo.flattenedMap["root"])
+						assert.Equal(t, expected, result)
+						if expected != result {
+							log.Printf("Test failed for operation: %v", operation)
+							allPassed = false
+						}
+					}
+				}
+			}
+			if !allPassed {
+				// print the matrix for debugging
+				log.Printf("Matrix:")
+				for k, v := range repo.flattenedMap {
+					log.Printf("Role: %s", k)
+					for k2, v2 := range v {
+						log.Printf("  Namespace: %s", k2)
+						for k3, v3 := range v2 {
+							log.Printf("    Resource: %s", k3)
+							for k4 := range v3 {
+								log.Printf("      Operation: %s", k4)
+							}
+						}
+					}
+				}
+				// print the role map for debugging
+				log.Printf("Role map:")
+				for k, v := range repo.RoleMap {
+					log.Printf("Role: %s", k)
+					log.Printf("  Subroles: %v", v.Subroles)
+					log.Printf("  Permit: %v", v.Permit)
+					log.Printf("  Deny: %v", v.Deny)
+				}
+				log.Printf("Subrole map:")
+				// make a list of [(subroleName, subrole)] from repo.SubroleMap, sort it by subroleName, and print it
+				subroleList := list.New()
+				for k, v := range repo.SubroleMap {
+					subroleList.PushBack(struct {
+						key   string
+						value *models.Role
+					}{k, v})
+				}
+				// Convert list to slice for sorting
+				subroleSlice := make([]struct {
+					key   string
+					value *models.Role
+				}, subroleList.Len())
+				i := 0
+				for e := subroleList.Front(); e != nil; e = e.Next() {
+					subroleSlice[i] = e.Value.(struct {
+						key   string
+						value *models.Role
+					})
+					i++
+				}
+				sort.Slice(subroleSlice, func(i, j int) bool {
+					return subroleSlice[i].key < subroleSlice[j].key
+				})
+				for _, subrole := range subroleSlice {
+					log.Printf("Subrole: %s", subrole.key)
+					log.Printf("  Subroles: %v", subrole.value.Subroles)
+					log.Printf("  Permit: %v", subrole.value.Permit)
+					log.Printf("  Deny: %v", subrole.value.Deny)
+				}
+			}
+		})
+	}
+}
+
+func createFuzzyRoleRepo(permits, denies, resources, namespaces int, pstar, dstar float64, operations []models.OperationType, graph map[string][]string) *RoleMapRepository {
```

Comment:
What is this? To make tests easier?

Reply:
This is generating a random role map, with given parameters. Its used for fuzzy tests - tests with random inputs. I decided this approach would be better than hardcoded test cases to avoid missing out any edge cases

Reply:
Understandable, have a great day